node/hyperbahn/service_proxy.js

```diff
@@ -257,6 +262,85 @@ function updateExitNodes(exitNodes, svcchan) {
     }
 };
 
+ServiceDispatchHandler.prototype.isBlocked =
+function isBlocked(cn, serviceName) {
+    var self = this;
+    if (!self.blockingTable) {
+        return false;
+    }
+
+    cn = cn || '*';
+    serviceName = serviceName || '*';
+
+    if (self.blockingTable[cn] && (
+        self.blockingTable[cn][serviceName] ||
+        self.blockingTable[cn]['*'])) {
+        return true;
+    } else if (self.blockingTable['*'] && (
+        self.blockingTable['*'][serviceName] ||
+        self.blockingTable['*']['*'])) {
+        return true;
+    }
+
+    return false;
+};
+
+ServiceDispatchHandler.prototype.block =
+function block(cn, serviceName) {
+    var self = this;
+    cn = cn || '*';
+    serviceName = serviceName || '*';
+
+    if (!self.blockingTable) {
+        self.blockingTable = {};
+    }
+
+    var target = self.blockingTable[cn];
+    if (!target) {
+        self.blockingTable[cn] = target = {};
```

Comment:
Nested objects are expensive. I recommend you do `blockTable[cn + '~~' + serviceName]`